### ISSUE TITLE
Better import conventions to avoid circular imports

### ISSUE THREAD

**Issue by msperber** [neulab/xnmt#439]
I have had problems with circular imports a number of times which can be annoying. More exhaustive type annotations (#438 ) might make this a more prominent issue.

It seems that most such issues can be avoided by importing modules instead of types from inside a module. E.g. instead of writing ```from xnmt.input import SimpleSentenceInput```, one would write ```import xnmt.input``` or something similar. This is already done in some places, but not consistently. See also https://stackoverflow.com/questions/7336802/how-to-avoid-circular-imports-in-python

There are 2 things to keep in mind:
- Doing so makes things more verbose. We probably want to use ```from xnmt import input``` / ```a=input.SimpleSentenceInput()``` instead of the more verbose ```import xnmt.import``` / ```a=xnmt.import.SimpleSentenceInput()```.
- A lot of the modules have unfortunate names. E.g. ```input``` is a reserved Python keyword, the ```batcher``` and ```decoder``` modules interfere with variables / arguments of the same name, etc. I would propose to rename the modules to more general names, e.g. ```batching```, ```inputs```, ```decoding```, etc.

**Comment by msperber** [comment]
Are there any comments / objections to the above suggestions? I'm getting an increasing number of import issues and would like to address this soon, though probably after #380 is merged to avoid merge conflicts there.

**Comment by philip30** [comment]
Hey sorry for the delay. I will be working on that soon.

**Comment by neubig** [comment]
I'm a little bit worried about the verbosity, but if there's no better way then I'm fine with this.

**Comment by msperber** [comment]
resolved by #550.
